packages/perseus/src/util/react-render.tsx

```diff
@@ -2,6 +2,23 @@ import {RenderStateRoot} from "@khanacademy/wonder-blocks-core";
 import * as React from "react";
 import * as ReactDOMClient from "react-dom/client";
 
+// We keep a cache of elements that have already had createRoot() called on
+// them. If we don't do this, React 18 warns every subsequent call:
+//   > You are calling ReactDOMClient.createRoot() on a container that has
+//   > already been passed to createRoot() before.
```

Comment:
Interesting... I'm surprised we haven't seen this in webapp. cc: @somewhatabstract

Reply:
It looks like webapp needed to do this also: https://github.com/Khan/webapp/blob/5133d22218437f9da3a7ef11bc89850b61378ac3/services/static/javascript/rendering/deprecated-react-render.tsx#L19